File: app/src/main/java/de/karzek/diettracker/presentation/main/diary/meal/GenericMealFragment.java
package de.karzek.diettracker.presentation.main.diary.meal;

import android.app.Activity;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentTransaction;
import android.support.v7.app.AppCompatDialogFragment;
import android.support.v7.widget.DividerItemDecoration;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.ViewStub;
import android.widget.FrameLayout;
import android.widget.TextView;

import java.util.ArrayList;
import java.util.HashMap;

import javax.inject.Inject;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.Unbinder;
import de.karzek.diettracker.R;
import de.karzek.diettracker.presentation.TrackerApplication;
import de.karzek.diettracker.presentation.common.BaseFragment;
import de.karzek.diettracker.presentation.main.diary.DiaryFragment;
import de.karzek.diettracker.presentation.main.diary.meal.adapter.diaryEntryList.DiaryEntryListAdapter;
import de.karzek.diettracker.presentation.main.diary.meal.adapter.favoriteRecipeList.FavoriteRecipeListAdapter;
import de.karzek.diettracker.presentation.main.diary.meal.dialog.MealSelectorDialog;
import de.karzek.diettracker.presentation.main.diary.meal.viewStub.CaloryDetailsView;
import de.karzek.diettracker.presentation.main.diary.meal.viewStub.CaloryMacroDetailsView;
import de.karzek.diettracker.presentation.model.DiaryEntryDisplayModel;
import de.karzek.diettracker.presentation.model.MealDisplayModel;
import de.karzek.diettracker.presentation.model.RecipeDisplayModel;
import de.karzek.diettracker.presentation.search.grocery.groceryDetail.GroceryDetailsActivity;
import de.karzek.diettracker.presentation.util.Constants;
import de.karzek.diettracker.presentation.util.StringUtils;

import static de.karzek.diettracker.presentation.util.SharedPreferencesUtil.VALUE_SETTING_NUTRITION_DETAILS_CALORIES_ONLY;

/**
 * Created by MarjanaKarzek on 28.05.2018.
 *
 * @author Marjana Karzek
 * @version 1.0
 * @date 28.05.2018
 */
public class GenericMealFragment extends BaseFragment implements GenericMealContract.View {

    @Override
    protected void setupFragmentComponent() {
        TrackerApplication.get(getContext()).createGenericMealComponent().inject(this);
    }

    @Inject
    GenericMealContract.Presenter presenter;

    @BindView(R.id.recycler_view_groceries)
    RecyclerView recyclerViewGroceries;
    @BindView(R.id.recycler_view_favorites)
    RecyclerView recyclerViewRecipes;
    @BindView(R.id.viewstub_calory_details)
    ViewStub caloryDetails;
    @BindView(R.id.viewstub_calory_macro_details)
    ViewStub caloryMacroDetails;
    @BindView(R.id.grocery_list_placeholder)
    TextView placeholder;
    @BindView(R.id.loading_view)
    FrameLayout loadingView;

    private Unbinder unbinder;

    OnRefreshViewPagerNeededListener callback;

    public interface OnRefreshViewPagerNeededListener {
        void onRefreshViewPagerNeeded();
    }

    private CaloryDetailsView detailsView;

    private String meal;
    private String selectedDate;

    private HashMap<String, Long> maxValues;

    @Override
    public void onAttach(Activity activity) {
        super.onAttach(activity);

        try {
            callback = (OnRefreshViewPagerNeededListener) activity;
        } catch (ClassCastException e) {
            throw new ClassCastException(activity.toString()
                    + " must implement OnRefreshViewPagerNeededListener");
        }
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.fragment_generic_meal, container, false);
        unbinder = ButterKnife.bind(this, view);

        meal = getArguments().getString("meal");
        selectedDate = ((DiaryFragment) getActivity().getSupportFragmentManager().findFragmentByTag("DiaryFragment")).getSelectedDate();

        return view;
    }

    @Override
    public void onViewCreated(View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        setupGroceryRecyclerView();
        setupRecipeRecyclerView();

        presenter.setView(this);
        presenter.setMeal(meal);

        presenter.start();
    }

    private void setupGroceryRecyclerView() {
        recyclerViewGroceries.setHasFixedSize(true);
        recyclerViewGroceries.setLayoutManager(new LinearLayoutManager(getContext()));
        recyclerViewGroceries.setAdapter(new DiaryEntryListAdapter(presenter, presenter, presenter, presenter));
        recyclerViewGroceries.addItemDecoration(new DividerItemDecoration(recyclerViewGroceries.getContext(),
                ((LinearLayoutManager) recyclerViewGroceries.getLayoutManager()).getOrientation()));
    }

    private void setupRecipeRecyclerView() {
        recyclerViewRecipes.setHasFixedSize(true);
        LinearLayoutManager manager = new LinearLayoutManager(getContext());
        manager.setOrientation(LinearLayoutManager.HORIZONTAL);
        recyclerViewRecipes.setLayoutManager(manager);
        recyclerViewRecipes.setAdapter(new FavoriteRecipeListAdapter(presenter));
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        presenter.finish();
        unbinder.unbind();
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        TrackerApplication.get(getContext()).releaseGenericMealComponent();
        callback = null;
    }

    @Override
    public void showGroceryListPlaceholder() {
        placeholder.setVisibility(View.VISIBLE);
    }

    @Override
    public void hideGroceryListPlaceholder() {
        placeholder.setVisibility(View.GONE);
    }

    @Override
    public void showNutritionDetails(String value) {
        if (value.equals(VALUE_SETTING_NUTRITION_DETAILS_CALORIES_ONLY)) {
            detailsView = new CaloryDetailsView(caloryDetails.inflate());
        } else {
            detailsView = new CaloryMacroDetailsView(caloryMacroDetails.inflate());
        }
    }

    @Override
    public void setNutritionMaxValues(HashMap<String, Long> values) {
        maxValues = values;
        detailsView.getCaloryProgressBarMaxValue().setText("" + values.get(Constants.CALORIES));

        if (detailsView instanceof CaloryMacroDetailsView) {
            ((CaloryMacroDetailsView) detailsView).getProteinProgressBarMaxValue().setText("von\n" + values.get(Constants.PROTEINS) + "g");
            ((CaloryMacroDetailsView) detailsView).getCarbsProgressBarMaxValue().setText("von\n" + values.get(Constants.CARBS) + "g");
            ((CaloryMacroDetailsView) detailsView).getFatsProgressBarMaxValue().setText("von\n" + values.get(Constants.FATS) + "g");
        }
    }

    @Override
    public void updateNutritionDetails(HashMap<String, Float> values) {
        detailsView.getCaloryProgressBar().setProgress(100.0f / maxValues.get(Constants.CALORIES) * values.get(Constants.CALORIES));
        detailsView.getCaloryProgressBarValue().setText("" + (int) values.get(Constants.CALORIES).floatValue());

        if (detailsView instanceof CaloryMacroDetailsView) {
            ((CaloryMacroDetailsView) detailsView).getProteinProgressBar().setProgress(100.0f / maxValues.get(Constants.PROTEINS) * values.get(Constants.PROTEINS));
            ((CaloryMacroDetailsView) detailsView).getProteinProgressBarValue().setText("" + StringUtils.formatFloat(values.get(Constants.PROTEINS)));

            ((CaloryMacroDetailsView) detailsView).getCarbsProgressBar().setProgress(100.0f / maxValues.get(Constants.CARBS) * values.get(Constants.CARBS));
            ((CaloryMacroDetailsView) detailsView).getCarbsProgressBarValue().setText("" + StringUtils.formatFloat(values.get(Constants.CARBS)));

            ((CaloryMacroDetailsView) detailsView).getFatsProgressBar().setProgress(100.0f / maxValues.get(Constants.FATS) * values.get(Constants.FATS));
            ((CaloryMacroDetailsView) detailsView).getFatsProgressBarValue().setText("" + StringUtils.formatFloat(values.get(Constants.FATS)));
        }
    }

    @Override
    public void updateGroceryList(ArrayList<DiaryEntryDisplayModel> diaryEntries) {
        ((DiaryEntryListAdapter) recyclerViewGroceries.getAdapter()).setList(diaryEntries);
    }

    @Override
    public void updateRecipeList(ArrayList<RecipeDisplayModel> recipes) {
        ((FavoriteRecipeListAdapter) recyclerViewRecipes.getAdapter()).setList(recipes);
    }

    @Override
    public String getSelectedDate() {
        return selectedDate;
    }

    @Override
    public void setSelectedDate(String date) {
        selectedDate = date;
        presenter.updateListItems(selectedDate);
    }

    @Override
    public void showLoading() {
        loadingView.setVisibility(View.VISIBLE);
    }

    @Override
    public void hideLoading() {
        loadingView.setVisibility(View.GONE);
    }

    @Override
    public void refreshRecyclerView() {
        presenter.updateListItems(selectedDate);
    }

    @Override
    public void showMoveDiaryEntryDialog(int id, ArrayList<MealDisplayModel> meals) {
        FragmentTransaction fragmentTransaction = getChildFragmentManager().beginTransaction();
        Fragment previous = getFragmentManager().findFragmentByTag("dialog");
        if (previous != null) {
            fragmentTransaction.remove(previous);
        }
        fragmentTransaction.addToBackStack(null);

        AppCompatDialogFragment dialogFragment = new MealSelectorDialog();
        Bundle bundle = new Bundle();
        bundle.putInt("id", id);
        bundle.putParcelableArrayList("meals", meals);
        dialogFragment.setArguments(bundle);
        dialogFragment.show(fragmentTransaction, "dialog");
    }

    @Override
    public void startEditMode(int id) {
        startActivity(GroceryDetailsActivity.newEditDiaryEntryIntent(getContext(), id));
    }

    @Override
    public void showRecipeRecyclerView() {
        recyclerViewRecipes.setVisibility(View.VISIBLE);
    }

    @Override
    public void hideRecipeRecyclerView() {
        recyclerViewRecipes.setVisibility(View.GONE);
    }

    @Override
    public void showGroceryRecyclerView() {
        recyclerViewGroceries.setVisibility(View.VISIBLE);
    }

    @Override
    public void hideGroceryRecyclerView() {
        recyclerViewGroceries.setVisibility(View.GONE);
    }

    @Override
    public void mealSelectedInDialog(int diaryEntryId, MealDisplayModel meal) {
        presenter.moveDiaryItemToMeal(diaryEntryId, meal);
        callback.onRefreshViewPagerNeeded();
    }
}

File: app/src/main/java/de/karzek/diettracker/presentation/main/diary/meal/GenericMealFragment.kt
package de.karzek.diettracker.presentation.main.diary.meal

import android.content.Context
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.recyclerview.widget.DividerItemDecoration
import androidx.recyclerview.widget.LinearLayoutManager
import de.karzek.diettracker.R
import de.karzek.diettracker.presentation.TrackerApplication
import de.karzek.diettracker.presentation.common.BaseFragment
import de.karzek.diettracker.presentation.main.diary.DiaryFragment
import de.karzek.diettracker.presentation.main.diary.meal.adapter.diaryEntryList.DiaryEntryListAdapter
import de.karzek.diettracker.presentation.main.diary.meal.adapter.favoriteRecipeList.FavoriteRecipeListAdapter
import de.karzek.diettracker.presentation.main.diary.meal.dialog.MealSelectorDialog
import de.karzek.diettracker.presentation.main.diary.meal.viewStub.CaloryDetailsView
import de.karzek.diettracker.presentation.main.diary.meal.viewStub.CaloryMacroDetailsView
import de.karzek.diettracker.presentation.model.DiaryEntryDisplayModel
import de.karzek.diettracker.presentation.model.MealDisplayModel
import de.karzek.diettracker.presentation.model.RecipeDisplayModel
import de.karzek.diettracker.presentation.search.grocery.groceryDetail.GroceryDetailsActivity
import de.karzek.diettracker.presentation.util.Constants
import de.karzek.diettracker.presentation.util.SharedPreferencesUtil.VALUE_SETTING_NUTRITION_DETAILS_CALORIES_ONLY
import de.karzek.diettracker.presentation.util.StringUtils
import kotlinx.android.synthetic.main.fragment_generic_meal.*
import kotlinx.android.synthetic.main.snippet_loading_view.*
import java.util.*
import javax.inject.Inject

/**
 * Created by MarjanaKarzek on 28.05.2018.
 *
 * @author Marjana Karzek
 * @version 1.0
 * @date 28.05.2018
 */
class GenericMealFragment : BaseFragment(), GenericMealContract.View {

    @Inject
    lateinit var presenter: GenericMealContract.Presenter

    private var callback: OnRefreshViewPagerNeededListener? = null

    private var detailsView: CaloryDetailsView? = null

    private var meal: String? = null
    private var selectedDate: String? = null

    private var maxValues: HashMap<String, Long>? = null

    override fun setupFragmentComponent() {
        TrackerApplication.get(context!!).createGenericMealComponent().inject(this)
    }

    interface OnRefreshViewPagerNeededListener {
        fun onRefreshViewPagerNeeded()
    }

    override fun onAttach(context: Context) {
        super.onAttach(context)

        try {
            callback = context as OnRefreshViewPagerNeededListener
        } catch (e: ClassCastException) {
            throw ClassCastException("$activity must implement OnRefreshViewPagerNeededListener")
        }

    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        val view = inflater.inflate(R.layout.fragment_generic_meal, container, false)

        meal = arguments!!.getString("meal")
        selectedDate = (activity!!.supportFragmentManager.findFragmentByTag("DiaryFragment") as DiaryFragment).selectedDate

        return view
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        setupGroceryRecyclerView()
        setupRecipeRecyclerView()

        presenter.setView(this)
        presenter.setMeal(meal)

        presenter.start()
    }

    private fun setupGroceryRecyclerView() {
        recycler_view_groceries.setHasFixedSize(true)
        recycler_view_groceries.layoutManager = LinearLayoutManager(context)
        recycler_view_groceries.adapter = DiaryEntryListAdapter(presenter, presenter, presenter, presenter)
        recycler_view_groceries.addItemDecoration(DividerItemDecoration(recycler_view_groceries.context,
                (recycler_view_groceries.layoutManager as LinearLayoutManager).orientation))
    }

    private fun setupRecipeRecyclerView() {
        recycler_view_favorites.setHasFixedSize(true)
        val manager = LinearLayoutManager(context)
        manager.orientation = LinearLayoutManager.HORIZONTAL
        recycler_view_favorites.layoutManager = manager
        recycler_view_favorites.adapter = FavoriteRecipeListAdapter(presenter)
    }

    override fun onDestroyView() {
        super.onDestroyView()
        presenter.finish()
    }

    override fun onDestroy() {
        super.onDestroy()
        TrackerApplication.get(context!!).releaseGenericMealComponent()
        callback = null
    }

    override fun showGroceryListPlaceholder() {
        grocery_list_placeholder.visibility = View.VISIBLE
    }

    override fun hideGroceryListPlaceholder() {
        grocery_list_placeholder.visibility = View.GONE
    }

    override fun showNutritionDetails(value: String) {
        detailsView = if (value == VALUE_SETTING_NUTRITION_DETAILS_CALORIES_ONLY) {
            CaloryDetailsView(viewstub_calory_details.inflate())
        } else {
            CaloryMacroDetailsView(viewstub_calory_macro_details.inflate())
        }
    }

    override fun setNutritionMaxValues(values: HashMap<String, Long>) {
        maxValues = values
        detailsView!!.caloryProgressBarMaxValue.text = "${values[Constants.CALORIES]!!}"

        if (detailsView is CaloryMacroDetailsView) {
            (detailsView as CaloryMacroDetailsView).proteinProgressBarMaxValue.text = "von\n${values[Constants.PROTEINS]}g"
            (detailsView as CaloryMacroDetailsView).carbsProgressBarMaxValue.text = "von\n${values[Constants.CARBS]}g"
            (detailsView as CaloryMacroDetailsView).fatsProgressBarMaxValue.text = "von\n${values[Constants.FATS]}g"
        }
    }

    override fun updateNutritionDetails(values: HashMap<String, Float>) {
        detailsView!!.caloryProgressBar.progress = 100.0f / maxValues!![Constants.CALORIES]!! * values[Constants.CALORIES]!!
        detailsView!!.caloryProgressBarValue.text = "${values[Constants.CALORIES]!!.toFloat().toInt()}"

        if (detailsView is CaloryMacroDetailsView) {
            (detailsView as CaloryMacroDetailsView).proteinProgressBar.progress = 100.0f / maxValues!![Constants.PROTEINS]!! * values[Constants.PROTEINS]!!
            (detailsView as CaloryMacroDetailsView).proteinProgressBarValue.text = StringUtils.formatFloat(values[Constants.PROTEINS]!!)

            (detailsView as CaloryMacroDetailsView).carbsProgressBar.progress = 100.0f / maxValues!![Constants.CARBS]!! * values[Constants.CARBS]!!
            (detailsView as CaloryMacroDetailsView).carbsProgressBarValue.text = StringUtils.formatFloat(values[Constants.CARBS]!!)

            (detailsView as CaloryMacroDetailsView).fatsProgressBar.progress = 100.0f / maxValues!![Constants.FATS]!! * values[Constants.FATS]!!
            (detailsView as CaloryMacroDetailsView).fatsProgressBarValue.text = StringUtils.formatFloat(values[Constants.FATS]!!)
        }
    }

    override fun updateGroceryList(diaryEntries: ArrayList<DiaryEntryDisplayModel>) {
        (recycler_view_groceries.adapter as DiaryEntryListAdapter).setList(diaryEntries)
    }

    override fun updateRecipeList(recipes: ArrayList<RecipeDisplayModel>) {
        (recycler_view_favorites.adapter as FavoriteRecipeListAdapter).setList(recipes)
    }

    override fun getSelectedDate(): String? {
        return selectedDate
    }

    override fun setSelectedDate(date: String) {
        selectedDate = date
        presenter.updateListItems(selectedDate)
    }

    override fun showLoading() {
        loading_view.visibility = View.VISIBLE
    }

    override fun hideLoading() {
        loading_view.visibility = View.GONE
    }

    override fun refreshRecyclerView() {
        presenter.updateListItems(selectedDate)
    }

    override fun showMoveDiaryEntryDialog(id: Int, meals: ArrayList<MealDisplayModel>) {
        val fragmentTransaction = childFragmentManager.beginTransaction()
        val previous = fragmentManager!!.findFragmentByTag("dialog")
        if (previous != null) {
            fragmentTransaction.remove(previous)
        }
        fragmentTransaction.addToBackStack(null)

        val dialogFragment = MealSelectorDialog()
        val bundle = Bundle()
        bundle.putInt("id", id)
        bundle.putParcelableArrayList("meals", meals)
        dialogFragment.arguments = bundle
        dialogFragment.show(fragmentTransaction, "dialog")
    }

    override fun startEditMode(id: Int) {
        startActivity(GroceryDetailsActivity.newEditDiaryEntryIntent(context!!, id))
    }

    override fun showRecipeRecyclerView() {
        recycler_view_favorites.visibility = View.VISIBLE
    }

    override fun hideRecipeRecyclerView() {
        recycler_view_favorites.visibility = View.GONE
    }

    override fun showGroceryRecyclerView() {
        recycler_view_groceries.visibility = View.VISIBLE
    }

    override fun hideGroceryRecyclerView() {
        recycler_view_groceries.visibility = View.GONE
    }

    override fun mealSelectedInDialog(diaryEntryId: Int, meal: MealDisplayModel) {
        presenter.moveDiaryItemToMeal(diaryEntryId, meal)
        callback!!.onRefreshViewPagerNeeded()
    }
}
